oxidation/crates/parsec_client_types/src/local_manifest.rs

```diff
@@ -55,6 +56,77 @@ pub struct Chunk {
     pub access: Option<BlockAccess>,
 }
 
+impl Chunk {
+    pub fn from_block_access(block_access: BlockAccess) -> Result<Self, &'static str> {
+        let raw_size = NonZeroU64::try_from(block_access.size).map_err(|_| "invalid raw size")?;
+        let stop = NonZeroU64::try_from(block_access.offset + block_access.size)
+            .map_err(|_| "invalid stop")?;
+        Ok(Self {
+            id: parsec_api_types::ChunkID::from(*block_access.id),
+            raw_offset: block_access.offset,
+            raw_size,
+            start: block_access.offset,
+            stop,
+            access: Some(block_access),
+        })
+    }
+
+    pub fn evolve_as_block(&self, data: &[u8]) -> Result<Self, &'static str> {
+        let mut result = self.clone();
+        // No-op
+        if result.is_block() {
+            return Ok(result);
+        }
+
+        // Check alignement
+        if result.raw_offset != result.start {
+            return Err("This chunk is not aligned");
+        }
+
+        // Craft access
+        result.access = Some(parsec_api_types::BlockAccess {
+            id: parsec_api_types::BlockID::from(*result.id),
+            key: parsec_api_crypto::SecretKey::generate(),
+            offset: result.start,
+            size: u64::from(result.stop) - result.start,
+            digest: parsec_api_crypto::HashDigest::from_data(data),
+        });
+
+        Ok(result)
+    }
+
+    pub fn is_block(&self) -> bool {
+        // Requires an access
+        if let Some(access) = &self.access {
+            // Pseudo block
+            !(!self.is_pseudo_block()
```

Comment:
double `!` brings poor readibility for me, I would prefer applying De Morgan's laws:
```rust
if let Some(access) = &self.access {
  // Pseudo block
  self.is_pseudo_block()
  // Offset inconsistent
  && self.raw_offset == access.offset
  // Size inconsistent
  && u64::from(self.raw_size) == access.size
} else {
  false
}
```

Reply:
Done :+1: